sql/src/test/java/io/crate/integrationtests/CastIntegrationTest.java

```diff
@@ -76,4 +76,26 @@ public void testInvalidCastExpression() throws Exception {
         expectedException.expectMessage("No cast function found for return type object");
         execute("select try_cast(name as array(object)) from sys.cluster");
     }
+
+    @Test
+    public void testImplicitCastColumnsWhenUsedOnBothSidesOfOperator() {
```

Comment:
I think a ExpressionAnalyzer / analyzer level unit test would be sufficient. We have already tests that verify that explicit casts work on an integration level.